Trust Anchor Groups (#104)
davidben/tls-trust-expressions: draft-ietf-tls-trust-anchor-ids.md

Reviewer: ```suggestion
The format defined in {{media-type}} can be used with ACME's alternate format mechanism (see {{Section 7.4.2 of !RFC8555}}) as follows. When downloading certificates, a supporting client SHOULD include "application/pem-certificate-chain-with-properties" in its HTTP Accept header ({{Section 12.5.1 of !RFC9110}}). When a supporting server sees such a header, it MAY then respond with that format to include a CertificatePropertyList with the certification path. This CertificatePropertyList MAY include `trust_anchor_id` and `trust_anchor_group_inclusions` properties for use with this protocol, or other properties defined in another document.
```

Author: Good catch. GitHub seems to have botched the diff, so I applied it by hand.

* When a relying party must update its policies to meet new security requirements, it adds to relying party diversity and the challenges that authenticating parties and CAs face. The relying party must then choose between compromising on user security or burdening the rest of the ecosystem, potentially impacting availability in the process.

To address this, this document introduces Trust Anchor Identifiers (Trust Anchor IDs). There are several parts to this mechanism:

1. {{trust-anchor-ids}} defines *trust anchor IDs*, which are short, unique identifiers for X.509 trust anchors, or groups of trust anchors.

2. {{tls-extension}} defines a TLS extension that communicates the relying party's requested trust anchors using trust anchor IDs. IDs that represent individual trust anchors can mitigate long X.509 names. IDs that represent groups of trust anchors can mitigate large lists of trust anchors.

> 3. {{retry-mechanism}} defines a retry mechanism that, when the relying party is a TLS client, can mitigate signaling failures. The server provides its list of available trust anchors in response so that the client can retry on mismatch. This can further mitigate large lists of trust anchors by giving deployments more flexibility in defining IDs for trust anchor groups. It also allows a TLS client to initially send only a subset of its trust anchors to the TLS server.

Reviewer 2: ```suggestion
3. {{retry-mechanism}} defines a retry mechanism that, when the relying party is a TLS client, can mitigate signaling failures. The server provides its list of available trust anchors in response so that the client can retry on mismatch. This can further mitigate large lists of trust anchors by giving deployments more flexibility in defining IDs for trust anchor groups, and it also allows a TLS client to initially send only a subset of its trust anchors to the TLS server, at the potential cost of a retry round trip in both cases. 
```

Author: Applied and wordsmithed a bit.

* For use in ASCII-compatible text protocols, a trust anchor ID's ASCII representation is the relative object identifier in dotted decimal notation. The example ID's ASCII representation is `32473.1`.

The length of a trust anchor ID's binary representation MUST NOT exceed 255 bytes. It SHOULD be significantly shorter, for bandwidth efficiency.

A trust anchor ID representing a single trust anchor SHOULD be allocated by the CA operator and be common among relying parties that trust the CA. They MAY be allocated by another party, e.g. when bootstrapping an existing ecosystem, if all parties agree on the ID. In particular, the protocol requires authenticating and relying parties to agree, and the authenticating party's configuration typically comes from the CA.

> A trust anchor ID representing a trust anchor group MAY be allocated by CA operators, relying party vendors, or other parties. {{trust-anchor-groups}} discusses defining trust anchor groups in more detail.

Reviewer 2: ```suggestion
A trust anchor ID representing a trust anchor group MAY be allocated by CA operators, relying party vendors, or other parties, but for them to be useful they require agreement between relying parties and authenticating parties. {{trust-anchor-groups}} discusses defining trust anchor groups in more detail.
```

## Trust Anchor Ranges

> Related trust anchor IDs can be allocated from a single OID arc. For example, a CA operator may allocate an OID arc, then append a version number, to describe snapshots of the CAs it operates. This section defines a *trust anchor range*, which describes a series of related trust anchor IDs. A trust anchor range is defined by three properties:

Reviewer 2: So.  Can this be a range of related groups of trust anchors? or only a range of related trust anchors?

Reviewer 2: because the language here says it's a range of "related trust anchor IDs" - but then goes on to define a "trust anchor range".

Reviewer 2: (as the original document defined Trust Anchor ID as well, an ID that uniquely identified a Trust Anchor, now morphing it to mean really a. "Trust Anchor Or Trust Group ID" means we may have some sharp edges here?)

Author: For better or worse, the groups are still identified by "trust anchor IDs" because that's describing the ID space. Slightly cursed but probably too much surgery to do now. But yeah let me improve this paragraph. I'm just trying to define a data structure in this section.

The following procedure can be used to perform this check. It succeeds if the range contains `id` and fails otherwise:

1. Check that `base` does not end in the middle of an OID component. That is, check that the most-significant bit of the last byte of `base` is unset. If it is set, fail the procedure.
2. Check that `base` is a prefix of `id`. If not, fail the procedure. Let `rest` be `id` with the `base` prefix removed.
3. Decode `rest` as a minimally-encoded, big-endian, base-128 OID component as follows:
   1. If `rest` is empty, fail the procedure.
   2. If the most-significant bit of the last byte of `rest` is set, fail the procedure.
   3. If the most-significant bit of any other byte of `rest` is unset, fail the procedure.
   4. If the first byte of `rest` is 0x80, fail the procedure.
   5. Set `v` to zero. Throughout this procedure, `v` will be less than 2<sup>64</sup>.
   6. For each byte `b` of `rest`:
      1. If `v` is greater than or equal to 2<sup>57</sup>, fail the procedure.
      2. Set `v` to `(v << 7) + (b & 127)`.
4. Check if `min <= v <= max`. If this is not true, fail the procedure. Otherwise, the procedure succeeds.

## Authenticating Party Configuration

> Authenticating parties are configured with one or more candidate certification paths to present in TLS, in some preference order. This preference order is used when multiple candidate paths are usable for a connection. For example, the authenticating party may prefer candidates that minimize message size or have more performant private keys.

Reviewer: Perhaps worth adding "certification path" to the terminology above? I also noticed a few instances of "certificate path" in the doc that should be normalized to "certification path". And some instances of "certificate authority" instead of "certification authority".

Author: Fixed the couple bad instances. (I think both of those terms are already in terminology section.)

Reviewer: Oops, yes "certification path" was already there!

Each candidate path that participates in this protocol must be configured with two properties:

> * The trust anchor ID for its corresponding trust anchor.

Reviewer: Does this have to correspond to a single trust anchor, or can it be a trust anchor group?

Author: I think this should be a single trust anchor. This is used when the server says "here are the CAs I have", etc., at which point we'd rather get an unambiguous signal. Otherwise the retry itself might have to pay for inaccuracies.

* Optionally, a list of *trust anchor group inclusions*. A trust anchor group inclusion is a trust anchor range ({{trust-anchor-ranges}}) that describes some trust anchor groups also containing the path's trust anchor.

> These properties allow certificate selection (see {{certificate-selection}}) to consider this path when a relying party advertises a matching trust anchor ID. It is RECOMMENDED, though not required, that this information come from the CA. {{certificate-properties}} defines a RECOMMENDED format for this information, along with an optional ACME {{!RFC8555}} extension for CAs to send it.

Reviewer 2: OK so here the information is "reccomended to come from the CA" (not "CA Operator" but "CA")

Later on we talk about a "defining party", which with appropriate irony is not defined, perhaps making this information?  We talk about including trust anchors in trust anchor ranges as defined in groups (which are in the same oid base as the TAI )  and then talk about "removing a CA".  Your comment above implies "here's the CA's I have" ;) so what are the implications of the above reccomendation for who is going to 
be the "defining party". 

If you want to use defining party (which I *think* means the holder of the base oid from which the trust anchor TAI's are assigned to trust anchors, and  therefore assigned to groups in the same oid base) I think it might be best to define it here, and then rationalize how "CA" "CA Operator" and "trust anchor" are being used here.

Author: Took out the weird "defining party" stuff. I think that was just confusing.

I think CA is correct here? The CA is the service that authenticating parties talk to. (Well, or the RA, or the ACME server, or whatever else. It's just a SHOULD.) A trust anchor is a CA that is configured in some relying party.

> Authenticating parties MAY have candidate certification paths without associated trust anchor IDs, but such paths will not participate in this protocol. Those paths MAY participate in other trust anchor negotiation protocols, such as the `certificate_authorities` extension.

Reviewer 2: And presumably, they may also have candidate certification paths now that do have associated trust anchor id's  which are agreed upon mutually by relying parties and authenticating parties, but may have partially or completely different views of trust anchor groups, and not agree there.

Reviewer 2: Meh, I think you pretty well got this in the follow on section, ponder if it needs at least a mention here? not sure it does now.

Author: Oh yeah "without associated trust anchor IDs" should be generalized. Fixed.

Relying parties MAY instead or additionally configure trust anchor IDs via some application-specific out-of-band information.

> Relying parties MAY additionally be configured with trust anchor groups that include its trust anchors. When authenticating parties are known to be configured with corresponding inclusion lists ({{authenticating-party-configuration}}), this can further reduce the size of messages sent by the relying party.

Reviewer 2: ```suggestion
Relying parties MAY additionally be configured with trust anchor groups that include their trust anchors. When authenticating parties are known to be configured with corresponding inclusion lists ({{authenticating-party-configuration}}), this can further reduce the size of messages sent by the relying party.
```

* The path's trust anchor ID appears in the relying party's `trust_anchors` extension.

> * One in the path's trust anchor group inclusions contains some ID in the relying party's `trust_anchors` extension.

Reviewer: typo?
```suggestion
* One of the path's trust anchor group inclusions contains some ID in the relying party's `trust_anchors` extension.
```

# DNS Service Parameter

> This section defines the `tls-trust-anchors` SvcParamKey {{!RFC9460}}. TLS servers can use this to advertise their available trust anchors in DNS, and aid the client in formulating its `trust_anchors` extension (see {{retry-mechanism}})ontg. This allows TLS deployments to support clients with many trust anchors without incurring the overhead of a reconnect.

Reviewer: ```suggestion
This section defines the `tls-trust-anchors` SvcParamKey {{!RFC9460}}. TLS servers can use this to advertise their available trust anchors in DNS, and aid the client in formulating its `trust_anchors` extension (see {{retry-mechanism}}). This allows TLS deployments to support clients with many trust anchors without incurring the overhead of a reconnect.
```

Author: Ahahaha what happened there? Thanks!

# Trust Anchor Groups

A trust anchor ID is typically much smaller than the corresponding X.509 name. Depending on the number of trust anchors, this can be sufficient to efficiently represent relying party state.

PKIs where further size savings are needed can use trust anchor groups ({{trust-anchor-ids}}). Trust anchor groups require additional coordination within a PKI, but they can further reduce relying party message sizes by allowing one ID to signal multiple trust anchors. To be usable, a trust anchor group must be known to relying parties (see {{relying-party-configuration}}) and configured with candidate paths in authenticating parties (see {{authenticating-party-configuration}}). The latter information SHOULD be provided by the CA, e.g., with the format defined in {{certificate-properties}}.

This section does not prescribe how to define trust anchor groups, but gives some general guidance:

A trust anchor group specifies a collection of trust anchors that are all trusted by some relying party. For example:

> * A CA operator might define a trust anchor group for the root CAs (or intermediate CAs, as described in {{intermediate-elision}}) that it expects supporting relying parties use as trust anchors.

Reviewer 2: is this define "a" trust anchor group, implying they'd only do one, or is this defining multiple trust anchor groups, perhaps with different relying parties using different groups which might in turn contain large or complete overlap of the same sets of trust anchors from the ca?

Author: Good point. The text immediately goes on to talk about versions, but this is confusion. I took another pass at this text. Does this help?

> * A relying party vendor might depend a trust anchor group for the trust anchors its relying parties trust.

Reviewer 2: "depend"? is this "depend on" or is this "define" - presumably they can't just define a rando one, only the "defining party / CA Operator" can assign a TAI that goes into a group in the same oid base. - right?

I think this needs a fix, or at least, I'm too dumb to not possibly draw wrong conclusions from it.

Author: Oops, yes, that indeed was supposed to be define. 😳

> Groups specific to a CA operator will cover fewer trust anchors, so a relying party might combine several operators' IDs to describe its trust anchors. However, such groups are more generally usable. A group specific to a relying party vendor can potentially be the only ID sent, but is less generally usable.

Reviewer 3: > However, such groups are more generally usable.

It is unclear what "such groups" refers to (could be "Groups specific to a CA operator" or the group created by a relying party.

Author: Ah yeah. For that matter, "more generally usable" is unclear because it doesn't say more generally usable than what. Let's maybe just delete that sentence and instead leave it to the following sentence.

> Groups specific to a CA operator will cover fewer trust anchors, so a relying party might combine several operators' IDs to describe its trust anchors. A group specific to a relying party vendor can potentially be the only ID sent, but is less generally usable.

Reviewer 2: The last part of this sentence and the hint that maybe a relying party could become a defining party? which seem like "no it would need to get the defining party / CA Operator to make a special group for it".  Why is that "less generally usable" - this seems a bold opinion without any real reason for it? 

If the result is say, I have three groups, one is every TAI the CA Operator feels like issueing from,  one is the TAI's that it can sign from that it thinks Chrome trusts, and one is the TAIs it can sign from that are chosen by weekly secret cabal of OpenBSD developers at the bar for OpenBSD use,  I'd argue that if I am a random relying party and don't want to run my *own* root program - wouldn't both those "not all the things the CA might have" groups possibly be more "generally useful" (I can tie my fate to Chrome's trust decisions, OpenBSD developers bar decisions, or the CA operators business decisions)

Again, I'm presuming here that the "defining party - probably the CA operator" is making the specific groups for each "relying party" - or to say "relying party with enough influence to get them to make a group for them"-  And if that's the case just call it what it is and call it a root program ;)  If that's the case I contend that identifiable groups for what's trusted by individual root programs are enormously "generally usable" with the flip side of "How do I get my own group" - what becomes the barrier to entry for a relying party - because it seems like this is contingent on me being able to get the owner of the OID arc issuing a group for me? Is that the case? if so I might have some other concerns...

Author: Hmm. I tried to crisp this up and found that talking about a "defining party" was just weird, since we don't _actually_ have anything for the defining party to _do_ in this document. I think that was just confusing, so I've rephrased it to not care about this. :-) Groups get defined somehow in the space of relying party <-> CA coordination, coordination that already happens as part of operating a root program.

Re less generally usable, I just mean that a group is useful in so far as relying parties tend to trust (almost) all of its contents as a unit. That's pretty solid for a single CA operator. As you get bigger, you're more and more making a judgement about how general it is that CAs A, B, and C are grouped together.

To that end, I meant for the "CA operator" vs "root program" to be more of a spectrum but didn't quite manage the first time. It doesn't have to be "CA operator" or "root program". It could be somewhere in between like "stuff common to this PKI". As you note, PKIs within a single space of applications (the web or email or whatever) are pretty uniform in practice, so I think there'll be plenty of room for good tradeoffs here.

In particular:

* If the group is slightly smaller than what you trust, send the group plus some other stuff. Maybe you send the extra CAs individually, or send a second group alongside it.
* If the group is slightly larger than what you trust, it's a little more awkward, though not necessarily fatal. We have the TAI retry flow.

> "How do I get my own group" - what becomes the barrier to entry for a relying party - because it seems like this is contingent on me being able to get the owner of the OID arc issuing a group for me? Is that the case? if so I might have some other concerns...

What did you mean by _the_ OID arc? If _you_ want to define a group, you define an OID from _your_ arc. Of course, you'll probably want to do that in coordination with the PKI you're part of, etc., so that CAs send that inclusion information. But how that looks is all a bit application-specific so the draft is kinda vague.

How's this new iteration look? I've tried to capture some of that in the text.

Over time, a CA operator may add or retire CAs, or relying parties may trust or distrust CAs. In this case, existing trust anchor groups SHOULD NOT be redefined, but the following versioning scheme MAY be used to reflect these and other changes:

> The defining party allocates an OID arc to define a series of related trust anchor groups. Each version of the group is identified by this OID arc, with an integer version number component appended. For example, a CA operator may define a trust anchor group series with the OID arc `32473.2`. Individual groups would have IDs `32473.2.0`, `32473.2.1`, `32473.2.2`, and so on.

Reviewer 3: > For example, a CA operator may define a trust anchor group series with the OID arc `32473.2`. Individual groups would have IDs `32473.2.0`, `32473.2.1`, `32473.2.2`, and so on.

Suggested rewording:

For example, if CA operator defines a trust anchor group series with the OID arc `32473.2`, the individual groups would have IDs `32473.2.0`, `32473.2.1`, `32473.2.2`, and so on.

Author: Done

> The defining party allocates an OID arc to define a series of related trust anchor groups. Each version of the group is identified by this OID arc, with an integer version number component appended. For example, if a CA operator defines a trust anchor group series with the OID arc `32473.2`, the individual groups would have IDs `32473.2.0`, `32473.2.1`, `32473.2.2`, and so on.

Reviewer 2: You need to define "defining party" - unless I've missed it.  Presumably the defining party is the party assigning the TAI from their OID arc that is associated with the trust anchor. now, ok it might now be
the CA operator itself, but certainly this information does have to get to the authenticating party. 
 
For that to work the defining party would nee to be stamping out TAI's on it's OID arc, and issuing groups on the same. If the intention here is that the defining party might not be the CA operator, but the CA operator is signing things that eventually get handed out to servers with the defining party's TAI on it, like, ok, but this also gets confusing with language saying "it should be the CA" (implying that the CA operator will also issue a *different* TAI and group on a different OID space?)

Reviewer 2: which I suppose the corollary is, as a server operator, do I expect to get four copies of the same chain signed by the same key, each with a different TAI assigned by a different "defining party" on a different OID base?

Author: > You need to define "defining party" - unless I've missed it.

From previous thread, I opted to just reduce dependence on "defining party" because it wasn't a terribly useful context here. 

> Presumably the defining party is the party assigning the TAI from their OID arc that is associated with the trust anchor. now, ok it might now be the CA operator itself, but certainly this information does have to get to the authenticating party.

The OID arc of the group does not need to have any relation to the OID of the individual trust anchors. There's nothing stopping group `Bob.1234` from containing CAs `David.42` and `David.5678`. But yes, to be useful, `David.42` and `David.5678` need to be somehow aware of `Bob.1234` to configure certs appropriately.

> which I suppose the corollary is, as a server operator, do I expect to get four copies of the same chain signed by the same key, each with a different TAI assigned by a different "defining party" on a different OID base?

Nah, one certificate chain with one CertificatePropertyList. Properties contain:
* A `trust_anchor_id` property containing the specific trust anchor. We still have TAIs that identify individual trust anchors and none of that changes. (For the retry, we really want to be able to say "oops, this group stuff got too fuzzy, I want _this one_.")
* A `trust_anchor_group_inclusions` with multiple inclusion records, one for each family of containing group.

> When describing trust anchor group inclusions, a CA can use a trust anchor range ({{trust-anchor-ranges}}) to describe which group versions include a candidate certification path's trust anchor:

Reviewer: I think we need something here indicating that this is metadata the CA passes to the authenticating party at issuance time, with a pointer to the CertificatePropertyList mechanism. It's sort of hinted at with the "at time of issuance" below, but not explicitly stated (or I missed it).

Author: Good call. I've rephrased this section to both clarify this and make it less insistent that the CA does this. (_Maybe_ you configure it manually. But also don't do that.)

* If the trust anchor is no longer in the latest version of the group, the range's `min` and `max` values are the first and last version that include the trust anchor, respectively.

* If the trust anchor is in the latest version of the group, at the time of issuance, the range's `min` value is the first version that includes the trust anchor, and `max` value is 2<sup>64</sup>-1.

In the second case, if the trust anchor is removed from later versions, the unlimited upper bound will become incorrect. The authenticating party TLS software may then misinterpret a relying party advertising a later version as supporting this trust anchor. Such signaling errors may result in the wrong certification path being selected. This can be mitigated in several ways:

> * If the authenticating party prefers a certificate from some replacement CA over the certificate from the removed CA, the correct certificate will still be chosen.

Reviewer 2: Ok, switching from removing a trust anchor, to removing a CA and preferring a replacement CA here is confusing - because above the CA is the thing you reccomended is giving me the tai and group informaiton. Please use the same language throughout, I suggest "trust anchor".  if the CA operator == one oid base, and I've decided I don't like the CA operator at all any more, all groups in that oid base go away. 

If I am rather removing one or more individual trust anchors then yes, those TAI's get a last version in their
included groups.  if the authenticating party then prefers a later or different TAI by virtue of a more preferred *TAI or Group* being sent, the correct certificate will be chosen? right?

Author: > Ok, switching from removing a trust anchor, to removing a CA and preferring a replacement CA here is confusing - because above the CA is the thing you reccomended is giving me the tai and group informaiton. Please use the same language throughout, I suggest "trust anchor".

Good call. Reworded. Is this version clearer?

> if the CA operator == one oid base, and I've decided I don't like the CA operator at all any more, all groups in that oid base go away.

This isn't talking about removing groups. This is talking about:

* RP trusts CAs A, B, and C.
* Group MyFancyGroup.10 contains CAs A, B, and C, so RP sends that.
* RP no longer trusts A
* MyFancyGroup.11 is defined that contains B and C, so RP sends that instead.
* Oops, existing certificates from A predate MyFancyGroup.11 and say "MyFancyGroup.{10..max}". Those certificates will incorrectly read as matching MyFancyGroup.11 and get sent when they shouldn't.

And then this list is talking about the ways that mitigate it. In particular, this bullet point is saying: if you have CAs A and B, and A got distrusted, getting an accurate read on A only matters if the server's preference order was A > B. But if A got distrusted and B is the replacement, presumably B > A is more likely. (Not guaranteed, but more likely.)

# Trust Anchor Groups

A trust anchor ID is typically much smaller than the corresponding X.509 name. Depending on the number of trust anchors, this can be sufficient to efficiently represent relying party state.

PKIs where further size savings are needed can use trust anchor groups ({{trust-anchor-ids}}). Trust anchor groups require additional coordination within a PKI, but they can further reduce relying party message sizes by allowing one ID to signal multiple trust anchors. To be usable, a trust anchor group must:

* be known to and sent by relying parties (see {{relying-party-configuration}}); and
* configured with candidate paths in authenticating parties (see {{authenticating-party-configuration}}), ideally provided by the CA during issuance (see {{certificate-properties}}).

This document does not prescribe how to define trust anchor groups, but gives some general guidance:

A trust anchor group specifies a collection of trust anchors, which a relying party can send to represent the contents. For example:

* A set of root CAs (or intermediate CAs, as in {{intermediate-elision}}) operated by a CA operator.

> * A set of trust anchors common to large set of relying parties. For example, trust anchors common to up-to-date web browsers, or common to some relying party vendor.

Reviewer 2: ```suggestion
* A set of trust anchors common to large set of relying parties. For example, trust anchors common to multiple root programs in an ecosystem
```

Author: (Ended up just deleting this sentence because it's already expanded on below.)

* A set of related application-specific trust anchors, such as a range of Merkle Tree Certificate landmarks {{?I-D.ietf-plants-merkle-tree-certs}}.

Different group definitions trade off size savings, applicability, and coordination overhead. A group that reflects a single CA operator will cover fewer trust anchors, so a relying party might combine several operators' IDs to describe its trust anchors. However, it is generally usable by relying parties that trust this CA operator. Such a group also requires minimal coordination for the CA operator to provide group inclusion information ({{authenticating-party-configuration}}) with the certificate.

> Conversely, a group that reflects a single relying party vendor can potentially be the only ID sent. However, it may be less generally usable when relying parties differ. Groups reflecting multiple relying party vendors are more broadly usable, but may need to be combined with other IDs in a given relying party. For example, a relying party might send a group containing established CAs common to its ecosystem, and individual IDs for its remaining, not yet as common CAs.

Reviewer 2: I really like this language, possibly consider my "root program" wording above, and tweak a little but I think this gets the right intention across.

Author: I think I prefer this version just because we haven't defined "root program" or anything. Actually maybe we should just delete that sentence. We already elaborate on it here.

This mechanism also allows servers to safely send fallback certificates that may not be as ubiquitously acceptable. Without some form of trust anchor negotiation, servers are limited to selecting certification paths that are ubiquitously trusted in all supported clients. This often means sending extra cross-certificates to target the lowest common denominator at a bandwidth cost. If the ClientHello contains `trust_anchors`, the server MAY opportunistically send a less ubiquitous, more bandwidth-efficient path based on local heuristics, with the expectation that the client will retry when the heuristics fail.

# Trust Anchor Groups

A trust anchor ID is typically much smaller than the corresponding X.509 name. Depending on the number of trust anchors, this can be sufficient to efficiently represent relying party state.

> PKIs where further size savings are needed can use trust anchor groups ({{trust-anchor-ids}}). Trust anchor groups can require additional coordination within a PKI. To be usable, a trust anchor group must be known to relying parties (see {{relying-party-configuration}}) and configured in with candidate paths in authenticating parties (see {{authenticating-party-configuration}}). However, they can further reduce relying party message sizes by allowing one ID to signal multiple trust anchors.

Reviewer: ```suggestion
PKIs where further size savings are needed can use trust anchor groups ({{trust-anchor-ids}}). Trust anchor groups require additional coordination within a PKI. To be usable, a trust anchor group must be known to relying parties (see {{relying-party-configuration}}) and configured with candidate paths in authenticating parties (see {{authenticating-party-configuration}}). However, they can further reduce relying party message sizes by allowing one ID to signal multiple trust anchors.
```

Over time, a CA operator may add or retire CAs, or relying parties may trust or distrust CAs. In this case, existing trust anchor groups SHOULD NOT be redefined, but the following versioning scheme MAY be used to reflect these and other changes:

The defining party allocates an OID arc to define a series of related trust anchor groups. Each version of the group is identified by this OID arc, with an integer version number component appended. For example, a CA operator may define a trust anchor group series with the OID arc `32473.2`. Individual groups would have IDs `32473.2.0`, `32473.2.1`, `32473.2.2`, and so on.

When describing trust anchor group inclusions, a CA can use a trust anchor range ({{trust-anchor-ranges}}) to describe which group versions include a candidate certification path's trust anchor:

* If the trust anchor is no longer in the latest version of the group, the CA sets `min` and `max` to the first and last version which include the trust anchor, respectively.

* If the trust anchor is in the latest version of the group, at the time of issuance, the CA sets `min` to the first version which includes the trust anchor, and `max` to 2<sup>64</sup>-1.

> In the second case, if the trust anchor is removed from later versions, the unlimited upper bound will become incorrect. The TLS software may then misinterpret a relying party advertising a later version as supporting this trust anchor. Such signaling errors may result in the wrong certification path being selected. This can be mitigated in several ways:

Reviewer: ```suggestion
In the second case, if the trust anchor is removed from later versions, the unlimited upper bound will become incorrect. The authenticating party TLS software may then misinterpret a relying party advertising a later version as supporting this trust anchor. Such signaling errors may result in the wrong certification path being selected. This can be mitigated in several ways:
```

The entries in a CertificatePropertyList MUST be sorted numerically by `type` and MUST NOT contain values with a duplicate `type`. Inputs that do not satisfy these invariants are syntax errors and MUST be rejected by parsers.

This document defines two properties:

* `trust_anchor_id`, defined in {{trust-anchor-id-property}}
* `additional_trust_anchor_ranges`, defined in {{additional-trust-anchor-ranges}}

Future documents MAY define other properties for use with other mechanisms. Such a document MUST define the format of the `data` field and how authenticating parties interpret the property. Authenticating parties MUST ignore properties with unrecognized CertificatePropertyType values.

## Trust Anchor ID Property

The `trust_anchor_id` property's `data` field contains the binary representation of the trust anchor ID of the certification path's trust anchor, as described in {{authenticating-party-configuration}}.

## Additional Trust Anchor Ranges

> The `additional_trust_anchor_ranges` property's `data` field contains the certification path's additional trust anchor ranges, as described in {{authenticating-party-configuration}}. The ranges are described in a TrustAnchorRangeList structure, defined below. Each TrustAnchorRange structure describes a trust anchor range, as defined in {{trust-anchor-ranges}}.

Reviewer: Any reason not to drop the `additional_` prefix and just call this `trust_anchor_ranges`?

Author: Could go either way. I used "additional" to capture that this doesn't change what is *the* trust anchor for the chain. These are just additional aliases. (Could also call it `trust_anchor_aliases`?)

Reviewer: Ah yeah makes sense! I kind of like `trust_anchor_aliases` or maybe `trust_anchor_alias_ranges`, but probably not enough to justify changing the existing text/name.

Author: Eh, changing it costs zero. :-) I'll send email soon so I guess people will be able to see this thread and chime in.

Author: (Resolving this since it's been renamed with the new "groups" framing.)
